lib/pdn/pdn.c

```diff
@@ -0,0 +1,763 @@
+		 * Pack the IP address family string.
+		 * Pack the APN gateway string.
+		 */
+
+		offset += snprintf(&p_command[offset],
+				   (PDN_AT_BUFFER_SIZE - offset), "%d,",
+				   handle->context_id);
+
+		const unsigned int family = handle->family;
+
+		if ((family & PDN_AF_INET6) && (family & PDN_AF_INET)) {
+			offset += snprintf(&p_command[offset],
+					   (PDN_AT_BUFFER_SIZE - offset),
+					   "\"IPV4V6\",");
+		} else if (family & PDN_AF_INET6) {
+			offset += snprintf(&p_command[offset],
+					   (PDN_AT_BUFFER_SIZE - offset),
+					   "\"IPV6\",");
+		} else {
+			offset += snprintf(&p_command[offset],
+					   (PDN_AT_BUFFER_SIZE - offset),
+					   "\"IP\",");
+		}
+
+		offset += snprintf(&p_command[offset],
+				   (PDN_AT_BUFFER_SIZE - offset), "\"%s\"",
+				   p_remote_addr);
+
+		if (offset > PDN_AT_BUFFER_SIZE) {
+			errno = E2BIG;
+			return -1;
+		}
+		break;
+	}
+	case CMD_PDN_ACTIVATE: {
+		/* Pack the context identifier. */
+		offset += snprintf(&p_command[offset],
+				   (PDN_AT_BUFFER_SIZE - offset), "%d",
+				   handle->context_id);
+		break;
+	}
+	case CMD_PDN_ID_GET: {
+		/* Pack the context identifier. */
+		offset += snprintf(&p_command[offset],
+				   (PDN_AT_BUFFER_SIZE - offset), "%d",
+				   handle->context_id);
+
+		resp_evt = EVT_PDN_ID_GET;
+		break;
+	}
+	case CMD_PDN_STATE_GET: {
+		/* No additional parameters. */
+		resp_evt = EVT_PDN_STATE_GET;
+		break;
+	}
+	case CMD_PDN_DEACTIVATE: {
+		/* Pack the context identifier. */
+		offset += snprintf(&p_command[offset],
+				   (PDN_AT_BUFFER_SIZE - offset), "%d",
+				   handle->context_id);
+		break;
+	}
+	case CMD_APN_DISCONNECT: {
+		/* Pack the context identifier. */
+		offset += snprintf(&p_command[offset],
+				   (PDN_AT_BUFFER_SIZE - offset), "%d",
+				   handle->context_id);
+		break;
+	}
+	default: {
+		/* not a valid command, return error */
+		errno = ENOTSUP;
+		return -1;
+	}
+	}
+
+	/* Send AT command and wait response */
+	err = at_cmd_write(p_command, p_command, PDN_AT_BUFFER_SIZE, NULL);
+	if (err != 0) {
+		free(p_command);
+		return -1;
+	}
+
+	/* Parse response */
+	err = command_response_parse(handle, p_command, resp_evt);
+	free(p_command);
+	return err;
+}
+
+/** @brief Function to remove all spaces from a given string */
+static void spaces_remove(char *const p_str)
+{
+	/* To keep track of non-space character count */
+	int count = 0;
+
+	/* Traverse the given string. */
+	for (int i = 0; p_str[i]; i++) {
+		/* If the current character is not a space,
+		 * then place it at index 'count'.
+		 */
+		if (p_str[i] != ' ') {
+			p_str[count] = p_str[i]; /* Increment the count. */
+			count++;
+		}
+	}
+	p_str[count] = '\0';
+}
+
+static int ok_rsp_parser(struct pdn_connection *const handle,
+			 char *const p_at_event)
+{
+	(void)handle;
+	(void)p_at_event;
+
+	return 0;
+}
+
+static int context_create_parser(struct pdn_connection *const handle,
+				 char *const p_at_event)
+{
+	spaces_remove(p_at_event);
```

Comment:
Seeing as atoi discards any leading whitespace characters, is this necessary?

> atoi. int atoi (const char * str); Convert string to integer. Parses the C-string str interpreting its content as an integral number, which is returned as a value of type int . The function first discards as many whitespace characters (as in isspace ) as necessary until the first non-whitespace character is found.

Reply:
Could the number have whitespace inside of it?

Reply:
True, that is not needed. I will remove the function.